johnsonandjohnson/Bodiless-JS
# feat(gatsby-theme-bodiless, backend): Better handling of merge conflicts

## Changes


*gatsby-theme-bodiless*
- Define pull message code for different use cases (A-F from original story).


*Backend*
- Include uncommitted file check for getChanges and getConflict, so user can review the conflict from local and upstream.
- Added copyfiles to @bodiless/backend for file operation.



## Test Instructions


- ### notes
 
  *Test has to be done on forked repo in order to test changes on origin master and upstream branch.*


1. #### Pre-test
    - fork the BodilessJS repo
    - checkout working branch: feat/better-handling-of-merge-conflicts-pr
    - make sure set origin to forked repo, master branch accept direct push.
    - create a test branch: feat/better-handling-of-merge-conflicts
    - add git remote upstream:
      `git remote add test https://github.com/dewen/Bodiless-JS.git`
1. #### Test case [A]: no production changes
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to './examples/test-site/src/data/pages/bulletpoints$default$bullet.json' or any other json content data files.`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "B1"`
    - `git push -f origin feat/better-handling-of-merge-conflicts`
1. #### Test [B] Changes are available but cannot be pulled, contact your development team for assistance.
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'package-lock.json'`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "A1"`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to './examples/test-site/src/data/pages/bulletpoints$default$bullet.json'`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "B1"`
    - `git push -f origin feat/better-handling-of-merge-conflicts`
1. #### Test [C] Pull success, your Edit Environment is up to date
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.json'`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "A1"`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `git push -f origin feat/better-handling-of-merge-conflicts`
1. #### Test [D] Your Edit Environment and Changeset differ from production on the following pages: page1name, page2name
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.json'`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "A1"`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.json' which has conflict with commit A1.`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "B1"`
    - `git push -f origin feat/better-handling-of-merge-conflicts`
    - **More cases include uncommitted changes on Edit env**
1. #### Test [E] Changes are available but cannot be pulled, contact your development team for assistance.
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.json'`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "A1"`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.json' which has conflict with commit A1.`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "B1"`
    - `git push -f origin feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - **More cases include uncommitted changes on Edit env**
1. #### Test [F] Changes are available but cannot be pulled, contact your development team for assistance
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.tsx'`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "A1"`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## Make some changes to 'examples/test-site/src/data/pages/**/*.json' which has conflict with commit A1.`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "B1"`
    - `git push -f origin feat/better-handling-of-merge-conflicts`
1. #### [G] An error has occurred, please try Pull again in a few minutes.
    - `git fetch test`
    - `git fetch origin`
    - `git stash`
    - `git checkout master`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `git push -f origin master`
    - `git checkout feat/better-handling-of-merge-conflicts`
    - `git reset --hard test/feat/better-handling-of-merge-conflicts-pr`
    - `## update packages/bodiless-backend/src/backend.js and remove getChange endpoint.`
    - `git add examples/test-site/src/data/pages/`
    - `git commit -m "B1"`
    - `git push -f origin feat/better-handling-of-merge-conflicts`


## Related Issues


Fixes #371
